llvm-general/test/LLVM/General/Test/Linking.hs

```diff
@@ -59,9 +59,9 @@ tests = testGroup "Linking" [
         ]      
 
     Module { moduleDefinitions = defs } <- withContext $ \context -> 
-      withModuleFromAST' context ast0 $ \m0 ->
-        withModuleFromAST' context ast1 $ \m1 -> do
-          runExceptT $ linkModules False m0 m1
-          moduleAST m0
+      withModuleFromAST' context ast0 $ \dest -> do
+        src <- failInIO $ astToFFIModule context ast1
+        failInIO $ linkModules dest src
```

Comment:
I don't think this direction with both ModuleRef and Module has worked out very well.
1) Both in fact represent references to mutable modules (just at different levels of indirection), so the distinction makes little sense from the outside.
2) It would in fact be quite common to create a ModuleRef and then want to call functions on the module before consuming it. That's how one would optimize before linking, for example.
3) Exposing getModule just puts us back where we were before this attempt.
4) As written, one can only link modules which come from ASTs, not from LLVMAssembly or bitcode. Solving that problem would in this approach require doubling all the various withModuleFrom... functions
5) You've made this attempt by duplicating functions, which is not ok in it's own right.
6) The overhead of using an IORef is not relevant in this context.

Please just make Module have an IORef.

I looked into the changes in LLVM which drove you here. The important part seems already to be present in 3.8, so perhaps these changes would be better made on that branch.